File: dpmp/link/discovery.py
# -*- coding: utf-8 -*-
"""局域网设备发现：UDP 广播 + 主动扫描 + 心跳。

节点按 device_id 去重（IP 变化不产生重复条目）。发现完全依赖
「请求 → 响应」：收到广播/扫描请求 → 回 reply；收到 reply → 登记节点。

端口：
  · 9998 UDP  广播 / 心跳
  · 9997 UDP  主动扫描（与收发隔离，避免扫描影响传输）
"""

import json
import socket
import threading
import time

from ..protocol import constants as C
from ..util.net import (get_all_local_ips, get_broadcast_addrs,
                        get_subnet_for_ip, get_mac_address)


class Node:
    """一个已发现的对端设备（dict 形式存储于 Discovery.nodes）。"""


class Discovery:
    """局域网设备发现。

    参数：
      device_id：本机稳定标识（用于去重，不显示自己）
      hostname：本机显示名
      log：日志回调
      on_new_node(ip, msg)：发现新节点时回调
      on_node_gone(ip)：节点下线时回调

    对外：
      nodes      {ip: node_dict}
      start() / stop()
      broadcast_search()        广播搜索
      scan_subnet(cidr)         扫描子网
      add_manual_node(ip)       手动添加
    """

    def __init__(self, device_id, hostname, log=None,
                 on_new_node=None, on_node_gone=None,
                 udp_port=None, scan_port=None, config=None):
        if config is None:
            from ..config import DEFAULT_CONFIG
            config = DEFAULT_CONFIG
        self.cfg = config
        self.discover_interval = config.discover_interval
        self.device_id = device_id or ""
        self.hostname = hostname or socket.gethostname()
        self.mac = get_mac_address()
        self.log = log or (lambda m: None)
        self.on_new_node = on_new_node
        self.on_node_gone = on_node_gone
        # 局域网端口（默认 9998 发现 / 9997 扫描，可自定义）
        self.udp_port = udp_port or C.LAN_UDP_PORT
        self.scan_port = scan_port or C.SCAN_PORT

        self.my_ips = get_all_local_ips(ipv6=False)
        self.my_ips_v6 = get_all_local_ips(ipv6=True)
        self.broadcast_addrs = get_broadcast_addrs()

        self.nodes = {}
        self.lock = threading.Lock()
        self.running = False
        self.broadcast_sockets = []
        self.listen_sockets = []
        self.auto_scan_enabled = False
        self.scanning = False
        self.pausing_network = False   # 传输时暂停广播/扫描/心跳

    # ---------- 生命周期 ----------

    def start(self):
        self.running = True
        threading.Thread(target=self._udp_listener, daemon=True).start()
        threading.Thread(target=self._scan_listener, daemon=True).start()
        self._start_broadcasters()
        self.log("本机 IPv4: %s" % ", ".join(self.my_ips))
        if self.my_ips_v6:
            self.log("本机 IPv6: %s" % ", ".join(self.my_ips_v6))
        self.log("广播地址: %s" % ", ".join(self.broadcast_addrs))

    def stop(self):
        self.running = False
        for s in self.broadcast_sockets + self.listen_sockets:
            try:
                s.close()
            except Exception:
                pass
        self._send_bye()

    # ---------- 消息构造 ----------

    def _build_msg(self, **extra):
        msg = {
            C.LAN_F_HOSTNAME: self.hostname,
            C.LAN_F_DEVICE_ID: self.device_id,
            C.LAN_F_MAC: self.mac,
        }
        msg.update(extra)
        return json.dumps(msg).encode("utf-8")

    def _upsert_node(self, remote_ip, msg, source="udp"):
        """按 device_id 去重地插入/更新节点。返回 True 表示是新节点。"""
        device_id = msg.get(C.LAN_F_DEVICE_ID, "") or ""
        hostname = msg.get(C.LAN_F_HOSTNAME, remote_ip) or remote_ip
        mac = msg.get(C.LAN_F_MAC, "") or ""
        is_new = False
        with self.lock:
            if device_id:
                for old_ip in list(self.nodes.keys()):
                    if old_ip == remote_ip:
                        continue
                    if self.nodes[old_ip].get(C.LAN_F_DEVICE_ID) == device_id:
                        del self.nodes[old_ip]
            if remote_ip in self.nodes:
                self.nodes[remote_ip][C.LAN_F_HOSTNAME] = hostname
                self.nodes[remote_ip][C.LAN_F_DEVICE_ID] = device_id
                self.nodes[remote_ip][C.LAN_F_MAC] = mac
                self.nodes[remote_ip]["last_seen"] = time.time()
                self.nodes[remote_ip]["heartbeat_fail"] = 0
            else:
                self.nodes[remote_ip] = {
                    C.LAN_F_HOSTNAME: hostname,
                    C.LAN_F_DEVICE_ID: device_id,
                    C.LAN_F_MAC: mac,
                    "last_seen": time.time(),
                    "source": source,
                    "heartbeat_fail": 0,
                }
                is_new = True
        return is_new

    def get_nodes(self):
        with self.lock:
            return dict(self.nodes)

    # ---------- 广播 ----------

    def _start_broadcasters(self):
        for ip_str in self.my_ips:
            try:
                sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
                sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
                sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
                sock.bind((ip_str, 0))
                self.broadcast_sockets.append(sock)
                threading.Thread(target=self._udp_broadcaster_on_sock,
                                 args=(sock, ip_str), daemon=True).start()
            except Exception as e:
                self.log("[警告] 无法为 %s 创建广播 socket: %s" % (ip_str, e))
        try:
            global_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            global_sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
            global_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self.broadcast_sockets.append(global_sock)
            threading.Thread(target=self._udp_broadcaster_on_sock,
                             args=(global_sock, None), daemon=True).start()
        except Exception as e:
            self.log("[警告] 全局广播 socket 创建失败: %s" % e)

    def _udp_broadcaster_on_sock(self, sock, bind_ip):
        """周期性广播搜索：前 5 秒每 0.2s，之后每 20s。"""
        start = time.time()
        while self.running:
            if not self.pausing_network:
                try:
                    sock.sendto(self._build_msg(**{C.LAN_K_DISCOVERY: True}),
                                ("255.255.255.255", self.udp_port))
                except Exception:
                    pass
            elapsed = time.time() - start
            if elapsed < 5:
                time.sleep(0.2)
            else:
                time.sleep(self.discover_interval)

    def broadcast_search(self):
        """广播搜索：向所有广播地址爆发式发送探测包，等待设备回复。"""
        self.log("[广播搜索] 发送广播探测，等待设备回复...")
        msg = self._build_msg(**{C.LAN_K_DISCOVERY: True})
        for _ in range(3):
            for bcast_addr in self.broadcast_addrs:
                try:
                    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
                    s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
                    s.settimeout(1)
                    s.sendto(msg, (bcast_addr, self.udp_port))
                    s.close()
                except Exception:
                    pass
            time.sleep(0.2)
        self.log("[广播搜索] 完成")

    # ---------- 接收 ----------

    def _udp_listener(self):
        sock4 = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock4.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock4.bind(("", self.udp_port))
        sock4.settimeout(1.0)
        self.listen_sockets.append(sock4)
        try:
            sock6 = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
            sock6.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            sock6.bind(("::", self.udp_port))
            sock6.settimeout(1.0)
            self.listen_sockets.append(sock6)
        except Exception:
            sock6 = None

        def listen_sock(sock, is_ipv6=False):
            while self.running:
                try:
                    data, addr = sock.recvfrom(1024)
                    msg = json.loads(data.decode("utf-8"))
                    remote_ip = addr[0]
                    if remote_ip in self.my_ips or remote_ip in self.my_ips_v6:
                        continue
                    if remote_ip in ("127.0.0.1", "::1", "0.0.0.0"):
                        continue
                    if msg.get(C.LAN_F_DEVICE_ID) and msg.get(C.LAN_F_DEVICE_ID) == self.device_id:
                        continue
                    if msg.get(C.LAN_K_BYE):
                        with self.lock:
                            removed = remote_ip in self.nodes
                            self.nodes.pop(remote_ip, None)
                        if removed and self.on_node_gone:
                            try:
                                self.on_node_gone(remote_ip)
                            except Exception:
                                pass
                        continue
                    if msg.get(C.LAN_K_HEARTBEAT):
                        with self.lock:
                            if remote_ip in self.nodes:
                                self.nodes[remote_ip]["last_seen"] = time.time()
                        reply = self._build_msg(**{C.LAN_K_HEARTBEAT: True, C.LAN_K_ACK: True})
                        sock.sendto(reply, (remote_ip, self.udp_port))
                        continue
                    is_new = self._upsert_node(remote_ip, msg, source="udp")
                    if is_new:
                        hn = msg.get(C.LAN_F_HOSTNAME, remote_ip)
                        self.log("[发现] 新设备 %s (%s)" % (hn, remote_ip))
                        if self.on_new_node:
                            try:
                                self.on_new_node(remote_ip, msg)
                            except Exception:
                                pass
                    if not msg.get(C.LAN_K_REPLY):
                        sock.sendto(self._build_msg(**{C.LAN_K_REPLY: True}),
                                    (remote_ip, self.udp_port))
                except socket.timeout:
                    continue
                except OSError:
                    break
                except Exception:
                    pass
            try:
                sock.close()
            except Exception:
                pass

        threading.Thread(target=listen_sock, args=(sock4, False), daemon=True).start()
        if sock6:
            threading.Thread(target=listen_sock, args=(sock6, True), daemon=True).start()

    # ---------- 扫描 ----------

    def _scan_listener(self):
        """扫描端口监听：收到扫描探测则回 scan_reply。"""
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            sock.bind(("", self.scan_port))
            sock.settimeout(1.0)
            self.listen_sockets.append(sock)
        except Exception as e:
            self.log("[扫描] 监听失败: %s" % e)
            return
        while self.running:
            try:
                data, addr = sock.recvfrom(1024)
                msg = json.loads(data.decode("utf-8"))
                remote_ip = addr[0]
                if msg.get(C.LAN_F_DEVICE_ID) and msg.get(C.LAN_F_DEVICE_ID) == self.device_id:
                    continue
                if remote_ip in self.my_ips or remote_ip in ("127.0.0.1", "::1"):
                    continue
                sock.sendto(self._build_msg(**{C.LAN_K_SCAN_REPLY: True}),
                            (remote_ip, self.scan_port))
                is_new = self._upsert_node(remote_ip, msg, source="scan")
                if is_new and self.on_new_node:
                    try:
                        self.on_new_node(remote_ip, msg)
                    except Exception:
                        pass
            except socket.timeout:
                continue
            except OSError:
                break
            except Exception:
                pass
        try:
            sock.close()
        except Exception:
            pass

    def scan_subnet(self, network_cidr, callback=None):
        """扫描一个子网（IPv4），发现的对端通过 on_new_node 回调。"""
        import ipaddress
        try:
            net = ipaddress.IPv4Network(network_cidr, strict=False)
        except Exception:
            return
        self.scanning = True
        msg = self._build_msg(**{C.LAN_K_SCAN_REPLY: True})
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            sock.settimeout(0.3)
            for host in net.hosts():
                if not self.running:
                    break
                try:
                    sock.sendto(msg, (str(host), self.scan_port))
                except Exception:
                    pass
        except Exception as e:
            self.log("[扫描] 子网 %s 扫描异常: %s" % (network_cidr, e))
        finally:
            try:
                sock.close()
            except Exception:
                pass
            self.scanning = False

    def add_manual_node(self, ip, hostname=None):
        """手动添加节点，并主动发探测让对方也发现我们。"""
        with self.lock:
            self.nodes[ip] = {
                "hostname": hostname or ip,
                "last_seen": time.time(),
                "source": "manual",
                "heartbeat_fail": 0,
            }
        self._send_probe_to(ip)
        self.log("[手动添加] 已向 %s 发送探测" % ip)

    def _send_probe_to(self, ip):
        try:
            family = socket.AF_INET6 if ":" in ip else socket.AF_INET
            s = socket.socket(family, socket.SOCK_DGRAM)
            s.settimeout(2)
            s.sendto(self._build_msg(), (ip, self.udp_port))
            try:
                s.sendto(self._build_msg(**{C.LAN_K_SCAN_REPLY: True}), (ip, self.scan_port))
            except Exception:
                pass
            s.close()
        except Exception as e:
            self.log("[手动添加] 向 %s 发送探测失败: %s" % (ip, e))

    # ---------- 下线 ----------

    def _send_bye(self):
        try:
            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
            s.sendto(self._build_msg(**{C.LAN_K_BYE: True}), ("255.255.255.255", self.udp_port))
            s.close()
        except Exception:
            pass

File: kotlin/src/main/kotlin/io/dpmp/link/Discovery.kt
package io.dpmp.link

import io.dpmp.Config
import io.dpmp.protocol.C
import io.dpmp.protocol.Json
import io.dpmp.util.Net
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.SocketTimeoutException
import java.util.concurrent.locks.ReentrantLock

/**
 * 局域网设备发现：UDP 广播 + 主动扫描 + 心跳。
 *
 * 节点按 device_id 去重（IP 变化不产生重复条目）。
 *
 * 端口：9998 UDP 广播/心跳；9997 UDP 主动扫描。
 *
 * 与 Python 端 dpmp/link/discovery.py 对齐。
 */
class Discovery(
    deviceId: String,
    hostname: String? = null,
    private val log: (String) -> Unit = {},
    private val onNewNode: ((String, Map<String, Any?>) -> Unit)? = null,
    private val onNodeGone: ((String) -> Unit)? = null,
    udpPort: Int? = null,
    scanPort: Int? = null,
    private val config: Config = Config.DEFAULT
) {
    private val discoverInterval = config.discoverInterval
    val deviceId: String = deviceId
    val hostname: String = hostname ?: try { InetAddress.getLocalHost().hostName } catch (_: Exception) { "unknown" }
    private val mac: String = Net.getMacAddress()

    val udpPort: Int = udpPort ?: C.LAN_UDP_PORT
    val scanPort: Int = scanPort ?: C.SCAN_PORT

    val myIps: List<String> = Net.getAllLocalIps(false)
    val myIpsV6: List<String> = Net.getAllLocalIps(true)
    private val broadcastAddrs: List<String> = Net.getBroadcastAddrs()

    private val nodes = HashMap<String, MutableMap<String, Any?>>()
    private val lock = ReentrantLock()
    @Volatile private var running = false
    private val broadcastSockets = ArrayList<DatagramSocket>()
    private val listenSockets = ArrayList<DatagramSocket>()
    @Volatile var autoScanEnabled = false
    @Volatile var scanning = false
    @Volatile var pausingNetwork = false

    // ---------- 生命周期 ----------

    fun start() {
        running = true
        Thread({ udpListener() }, "dpmp-disc-udp").apply { isDaemon = true; start() }
        Thread({ scanListener() }, "dpmp-disc-scan").apply { isDaemon = true; start() }
        startBroadcasters()
        log("本机 IPv4: " + myIps.joinToString(", "))
        if (myIpsV6.isNotEmpty()) log("本机 IPv6: " + myIpsV6.joinToString(", "))
        log("广播地址: " + broadcastAddrs.joinToString(", "))
    }

    fun stop() {
        running = false
        for (s in broadcastSockets + listenSockets) {
            try { s.close() } catch (_: Exception) {}
        }
        sendBye()
    }

    // ---------- 消息构造 ----------

    private fun buildMsg(vararg extra: Pair<String, Any?>): ByteArray {
        val msg = LinkedHashMap<String, Any?>()
        msg[C.LAN_F_HOSTNAME] = hostname
        msg[C.LAN_F_DEVICE_ID] = deviceId
        msg[C.LAN_F_MAC] = mac
        for ((k, v) in extra) msg[k] = v
        return Json.encode(msg).toByteArray(Charsets.UTF_8)
    }

    /** 按 device_id 去重地插入/更新节点。返回 true 表示新节点。 */
    private fun upsertNode(remoteIp: String, msg: Map<String, Any?>, source: String = "udp"): Boolean {
        val did = msg[C.LAN_F_DEVICE_ID] as? String ?: ""
        val hn = msg[C.LAN_F_HOSTNAME] as? String ?: remoteIp
        val m = msg[C.LAN_F_MAC] as? String ?: ""
        var isNew = false
        lock.lock()
        try {
            if (did.isNotEmpty()) {
                for (oldIp in nodes.keys.toList()) {
                    if (oldIp == remoteIp) continue
                    if (nodes[oldIp]?.get(C.LAN_F_DEVICE_ID) == did) nodes.remove(oldIp)
                }
            }
            val existing = nodes[remoteIp]
            if (existing != null) {
                existing[C.LAN_F_HOSTNAME] = hn
                existing[C.LAN_F_DEVICE_ID] = did
                existing[C.LAN_F_MAC] = m
                existing["last_seen"] = nowSec()
                existing["heartbeat_fail"] = 0
            } else {
                val n = HashMap<String, Any?>()
                n[C.LAN_F_HOSTNAME] = hn; n[C.LAN_F_DEVICE_ID] = did; n[C.LAN_F_MAC] = m
                n["last_seen"] = nowSec(); n["source"] = source; n["heartbeat_fail"] = 0
                nodes[remoteIp] = n
                isNew = true
            }
        } finally {
            lock.unlock()
        }
        return isNew
    }

    fun getNodes(): Map<String, Map<String, Any?>> {
        lock.lock(); try { return nodes.mapValues { HashMap(it.value) } } finally { lock.unlock() }
    }

    // ---------- 广播 ----------

    private fun startBroadcasters() {
        for (ipStr in myIps) {
            try {
                val s = DatagramSocket(null)
                s.broadcast = true
                s.reuseAddress = true
                s.bind(InetSocketAddress(ipStr, 0))
                broadcastSockets.add(s)
                Thread({ udpBroadcasterOnSock(s) }, "dpmp-disc-bcast").apply { isDaemon = true; start() }
            } catch (e: Exception) {
                log("[警告] 无法为 " + ipStr + " 创建广播 socket: " + e.message)
            }
        }
        try {
            val g = DatagramSocket(null)
            g.broadcast = true
            g.reuseAddress = true
            broadcastSockets.add(g)
            Thread({ udpBroadcasterOnSock(g) }, "dpmp-disc-bcast-global").apply { isDaemon = true; start() }
        } catch (e: Exception) {
            log("[警告] 全局广播 socket 创建失败: " + e.message)
        }
    }

    private fun udpBroadcasterOnSock(sock: DatagramSocket) {
        val start = System.currentTimeMillis()
        while (running) {
            if (!pausingNetwork) {
                try {
                    val msg = buildMsg(C.LAN_K_DISCOVERY to true)
                    sock.send(DatagramPacket(msg, msg.size,
                        InetAddress.getByName("255.255.255.255"), udpPort))
                } catch (_: Exception) {}
            }
            val elapsed = (System.currentTimeMillis() - start) / 1000.0
            Thread.sleep(if (elapsed < 5) 200 else discoverInterval * 1000L)
        }
    }

    fun broadcastSearch() {
        log("[广播搜索] 发送广播探测，等待设备回复...")
        val msg = buildMsg(C.LAN_K_DISCOVERY to true)
        for (r in 0 until 3) {
            for (bcast in broadcastAddrs) {
                try {
                    val s = DatagramSocket()
                    s.broadcast = true
                    s.soTimeout = 1000
                    s.send(DatagramPacket(msg, msg.size, InetAddress.getByName(bcast), udpPort))
                    s.close()
                } catch (_: Exception) {}
            }
            Thread.sleep(200)
        }
        log("[广播搜索] 完成")
    }

    // ---------- 接收 ----------

    private fun udpListener() {
        val sock4: DatagramSocket
        try {
            sock4 = DatagramSocket(null)
            sock4.reuseAddress = true
            sock4.bind(InetSocketAddress(udpPort))
            sock4.soTimeout = 1000
            listenSockets.add(sock4)
        } catch (e: Exception) {
            log("[发现] UDP 监听失败: " + e.message); return
        }
        var sock6: DatagramSocket? = null
        try {
            sock6 = DatagramSocket(null)
            sock6.reuseAddress = true
            sock6.bind(InetSocketAddress("::", udpPort))
            sock6.soTimeout = 1000
            listenSockets.add(sock6)
        } catch (_: Exception) { sock6 = null }

        Thread({ listenSock(sock4) }, "dpmp-disc-listen4").apply { isDaemon = true; start() }
        sock6?.let { Thread({ listenSock(it) }, "dpmp-disc-listen6").apply { isDaemon = true; start() } }
    }

    private fun listenSock(sock: DatagramSocket) {
        val buf = ByteArray(1024)
        while (running) {
            try {
                val pkt = DatagramPacket(buf, buf.size)
                sock.receive(pkt)
                val msg = try { Json.decode(String(pkt.data, 0, pkt.length, Charsets.UTF_8)) } catch (_: Exception) { continue }
                val remoteIp = pkt.address.hostAddress
                if (remoteIp in myIps || remoteIp in myIpsV6) continue
                if (remoteIp == "127.0.0.1" || remoteIp == "::1" || remoteIp == "0.0.0.0") continue
                val did = msg[C.LAN_F_DEVICE_ID] as? String ?: ""
                if (did.isNotEmpty() && did == deviceId) continue
                if (msg[C.LAN_K_BYE] == true) {
                    val removed = lock.let { it.lock(); try { nodes.remove(remoteIp) != null } finally { it.unlock() } }
                    if (removed) try { onNodeGone?.invoke(remoteIp) } catch (_: Exception) {}
                    continue
                }
                if (msg[C.LAN_K_HEARTBEAT] == true) {
                    lock.lock(); try { nodes[remoteIp]?.put("last_seen", nowSec()) } finally { lock.unlock() }
                    val reply = buildMsg(C.LAN_K_HEARTBEAT to true, C.LAN_K_ACK to true)
                    sock.send(DatagramPacket(reply, reply.size, InetAddress.getByName(remoteIp), udpPort))
                    continue
                }
                val isNew = upsertNode(remoteIp, msg, "udp")
                if (isNew) {
                    val hn = msg[C.LAN_F_HOSTNAME] as? String ?: remoteIp
                    log("[发现] 新设备 " + hn + " (" + remoteIp + ")")
                    try { onNewNode?.invoke(remoteIp, msg) } catch (_: Exception) {}
                }
                if (msg[C.LAN_K_REPLY] != true) {
                    val reply = buildMsg(C.LAN_K_REPLY to true)
                    sock.send(DatagramPacket(reply, reply.size, InetAddress.getByName(remoteIp), udpPort))
                }
            } catch (_: SocketTimeoutException) {
                continue
            } catch (_: Exception) {
                break
            }
        }
        try { sock.close() } catch (_: Exception) {}
    }

    // ---------- 扫描 ----------

    private fun scanListener() {
        val sock: DatagramSocket
        try {
            sock = DatagramSocket(null)
            sock.reuseAddress = true
            sock.bind(InetSocketAddress(scanPort))
            sock.soTimeout = 1000
            listenSockets.add(sock)
        } catch (e: Exception) {
            log("[扫描] 监听失败: " + e.message); return
        }
        val buf = ByteArray(1024)
        while (running) {
            try {
                val pkt = DatagramPacket(buf, buf.size)
                sock.receive(pkt)
                val msg = try { Json.decode(String(pkt.data, 0, pkt.length, Charsets.UTF_8)) } catch (_: Exception) { continue }
                val remoteIp = pkt.address.hostAddress
                val did = msg[C.LAN_F_DEVICE_ID] as? String ?: ""
                if (did.isNotEmpty() && did == deviceId) continue
                if (remoteIp in myIps || remoteIp == "127.0.0.1" || remoteIp == "::1") continue
                val reply = buildMsg(C.LAN_K_SCAN_REPLY to true)
                sock.send(DatagramPacket(reply, reply.size, InetAddress.getByName(remoteIp), scanPort))
                val isNew = upsertNode(remoteIp, msg, "scan")
                if (isNew) try { onNewNode?.invoke(remoteIp, msg) } catch (_: Exception) {}
            } catch (_: SocketTimeoutException) {
                continue
            } catch (_: Exception) {
                break
            }
        }
        try { sock.close() } catch (_: Exception) {}
    }

    /** 扫描一个子网（IPv4），发现的对端通过 onNewNode 回调。 */
    fun scanSubnet(networkCidr: String, callback: ((String) -> Unit)? = null) {
        val range = cidrHosts(networkCidr) ?: return
        scanning = true
        val msg = buildMsg(C.LAN_K_SCAN_REPLY to true)
        var sock: DatagramSocket? = null
        try {
            sock = DatagramSocket()
            sock.soTimeout = 300
            for (host in range) {
                if (!running) break
                try {
                    sock.send(DatagramPacket(msg, msg.size, InetAddress.getByName(host), scanPort))
                } catch (_: Exception) {}
            }
            callback?.invoke(networkCidr)
        } catch (e: Exception) {
            log("[扫描] 子网 " + networkCidr + " 扫描异常: " + e.message)
        } finally {
            try { sock?.close() } catch (_: Exception) {}
            scanning = false
        }
    }

    /** 手动添加节点，并主动发探测让对方也发现我们。 */
    fun addManualNode(ip: String, hostname: String? = null) {
        lock.lock()
        try {
            val n = HashMap<String, Any?>()
            n["hostname"] = hostname ?: ip
            n["last_seen"] = nowSec(); n["source"] = "manual"; n["heartbeat_fail"] = 0
            nodes[ip] = n
        } finally {
            lock.unlock()
        }
        sendProbeTo(ip)
        log("[手动添加] 已向 " + ip + " 发送探测")
    }

    private fun sendProbeTo(ip: String) {
        try {
            val s = DatagramSocket()
            s.soTimeout = 2000
            val m1 = buildMsg()
            s.send(DatagramPacket(m1, m1.size, InetAddress.getByName(ip), udpPort))
            try {
                val m2 = buildMsg(C.LAN_K_SCAN_REPLY to true)
                s.send(DatagramPacket(m2, m2.size, InetAddress.getByName(ip), scanPort))
            } catch (_: Exception) {}
            s.close()
        } catch (e: Exception) {
            log("[手动添加] 向 " + ip + " 发送探测失败: " + e.message)
        }
    }

    private fun sendBye() {
        try {
            val s = DatagramSocket()
            s.broadcast = true
            val m = buildMsg(C.LAN_K_BYE to true)
            s.send(DatagramPacket(m, m.size, InetAddress.getByName("255.255.255.255"), udpPort))
            s.close()
        } catch (_: Exception) {}
    }

    companion object {
        private fun nowSec(): Double = System.nanoTime() / 1_000_000_000.0

        /** 生成 CIDR 内的所有主机 IP（不含网络号与广播地址）。 */
        private fun cidrHosts(cidr: String): List<String>? {
            val slash = cidr.indexOf('/')
            if (slash < 0) return null
            val ip = cidr.substring(0, slash)
            val prefix = cidr.substring(slash + 1).toIntOrNull() ?: return null
            val octets = ip.split(".").map { it.toIntOrNull() ?: return null }
            if (octets.size != 4) return null
            var addr = 0L
            for (o in octets) addr = (addr shl 8) or o.toLong()
            val mask = if (prefix == 0) 0L else (0xFFFFFFFFL shl (32 - prefix)) and 0xFFFFFFFFL
            val network = addr and mask
            val broadcast = network or (mask.inv() and 0xFFFFFFFFL)
            val out = ArrayList<String>()
            var cur = network + 1
            var count = 0
            while (cur < broadcast && count < 65536) {
                out.add(listOf(
                    (cur shr 24) and 0xFF, (cur shr 16) and 0xFF,
                    (cur shr 8) and 0xFF, cur and 0xFF
                ).joinToString("."))
                cur++
                count++
            }
            return out
        }
    }
}
